src/System.Device.Gpio/System/Device/Media/Devices/UnixVideoDevice.Linux.cs

```diff
@@ -0,0 +1,420 @@
+                memory = v4l2_memory.V4L2_MEMORY_MMAP
+            };
+            V4l2Struct(VideoSettings.VIDIOC_REQBUFS, ref req);
+
+            // Mapping the applied buffer to user space
+            V4l2FrameBuffer* buffers = stackalloc V4l2FrameBuffer[4];
+            for (uint i = 0; i < BufferCount; i++)
+            {
+                v4l2_buffer buffer = new v4l2_buffer
+                {
+                    index = i,
+                    type = v4l2_buf_type.V4L2_BUF_TYPE_VIDEO_CAPTURE,
+                    memory = v4l2_memory.V4L2_MEMORY_MMAP
+                };
+                V4l2Struct(VideoSettings.VIDIOC_QUERYBUF, ref buffer);
+
+                buffers[i].Length = buffer.length;
+                buffers[i].Start = Interop.mmap(IntPtr.Zero, (int)buffer.length, MemoryMappedProtections.PROT_READ | MemoryMappedProtections.PROT_WRITE, MemoryMappedFlags.MAP_SHARED, _deviceFileDescriptor, (int)buffer.m.offset);
+            }
+
+            // Put the buffer in the processing queue
+            for (uint i = 0; i < BufferCount; i++)
+            {
+                v4l2_buffer buffer = new v4l2_buffer
+                {
+                    index = i,
+                    type = v4l2_buf_type.V4L2_BUF_TYPE_VIDEO_CAPTURE,
+                    memory = v4l2_memory.V4L2_MEMORY_MMAP
+                };
+                V4l2Struct(VideoSettings.VIDIOC_QBUF, ref buffer);
+            }
+
+            // Start data stream
+            v4l2_buf_type type = v4l2_buf_type.V4L2_BUF_TYPE_VIDEO_CAPTURE;
+            Interop.ioctl(_deviceFileDescriptor, (int)VideoSettings.VIDIOC_STREAMON, new IntPtr(&type));
+
+            // Get one frame from the buffer
+            v4l2_buffer frame = new v4l2_buffer
+            {
+                type = v4l2_buf_type.V4L2_BUF_TYPE_VIDEO_CAPTURE,
+                memory = v4l2_memory.V4L2_MEMORY_MMAP,
+            };
+            V4l2Struct(VideoSettings.VIDIOC_DQBUF, ref frame);
+
+            // Get data from pointer
+            IntPtr intptr = buffers[frame.index].Start;
+            byte[] dataBuffer = new byte[buffers[frame.index].Length];
+            Marshal.Copy(source: intptr, destination: dataBuffer, startIndex: 0, length: (int)buffers[frame.index].Length);
+
+            // Requeue the buffer
+            V4l2Struct(VideoSettings.VIDIOC_QBUF, ref frame);
+
+            // Close data stream
+            Interop.ioctl(_deviceFileDescriptor, (int)VideoSettings.VIDIOC_STREAMOFF, new IntPtr(&type));
+
+            // Unmapping the applied buffer to user space
+            for (uint i = 0; i < BufferCount; i++)
+            {
+                Interop.munmap(buffers[i].Start, (int)buffers[i].Length);
+            }
+
+            return dataBuffer;
+        }
+
+        private unsafe void SetVideoConnectionSettings()
+        {
+            FillVideoConnectionSettings();
+
+            // Set capture format
+            v4l2_format format;
+            if (Settings.CaptureSize == (0, 0))
```

Comment:
My Raspberry Pi CSI camera(unofficial) couldn't get maximum resolution, but the USB camera could get it. I guess it's a driver problem.

Reply:
Did you try changing other parameters while increasing the size? (i.e. slow down capture rate or something similar - might be worth to check with strace what other software is doing)